<a href="https://colab.research.google.com/github/UadrianSP/am-turma5/blob/main/TratamentoCensus_aula0901.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pipeline de Analise de Dados [Census]

## Importação de Bibliotecas

*   Pandas
*   Numpy
*   MatPlotLib
*   SeaBorn



In [46]:
!pip install pandas
!pip install numpy
!pip install matplotlib
!pip install seaborn

In [47]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [48]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [49]:
import os

def localizar_arquivo(nome_arquivo, pasta_inicial='/content/drive'):
    print(f"Buscando por '{nome_arquivo}' em {pasta_inicial}...")
    for raiz, dirs, arquivos in os.walk(pasta_inicial):
        if nome_arquivo in arquivos:
            caminho_completo = os.path.join(raiz, nome_arquivo)
            print(f"\n✅ ARQUIVO ENCONTRADO! Caminho correto:")
            print(caminho_completo)
            return caminho_completo
    print(f"\n Arquivo '{nome_arquivo}' não encontrado no Drive.")
    return None

caminho_real = localizar_arquivo('census.csv')

Buscando por 'census.csv' em /content/drive...

✅ ARQUIVO ENCONTRADO! Caminho correto:
/content/drive/MyDrive/am-t6/census.csv


In [50]:
import os

# Caminho que você está tentando acessar
caminho_alvo = '/content/drive/MyDrive/am-t6/'

if os.path.exists(caminho_alvo):
    print(f"Arquivos encontrados em {caminho_alvo}:")
    print(os.listdir(caminho_alvo))
else:
    print(f"A pasta {caminho_alvo} não foi encontrada. Listando o conteúdo de 'MyDrive' para ajudar:")
    print(os.listdir('/content/drive/MyDrive/'))

Arquivos encontrados em /content/drive/MyDrive/am-t6/:
['census.csv', 'credit_data.csv']


In [51]:
import os

# Vamos verificar a raiz do drive montado
raiz_drive = '/content/drive'

if os.path.exists(raiz_drive):
    print(f"Pastas encontradas em {raiz_drive}:")
    try:
        print(os.listdir(raiz_drive))
        # Tenta encontrar o arquivo recursivamente se a lista acima não ajudar
        print("\nBuscando 'census.csv' em subpastas...")
        for root, dirs, files in os.walk(raiz_drive):
            if 'census.csv' in files:
                print(f"ARQUIVO ENCONTRADO EM: {os.path.join(root, 'census.csv')}")
    except Exception as e:
        print(f"Erro ao acessar subpastas: {e}")
else:
    print("O drive não parece estar montado em /content/drive. Tente rodar a célula de mount novamente.")

Pastas encontradas em /content/drive:
['.shortcut-targets-by-id', 'MyDrive', '.Trash-0', '.Encrypted']

Buscando 'census.csv' em subpastas...
ARQUIVO ENCONTRADO EM: /content/drive/MyDrive/am-t6/census.csv


## Carregamento dos Dados da Base

In [52]:
base_census = pd.read_csv('/content/drive/MyDrive/am-t6/census.csv')

In [53]:
base_census

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


In [54]:
base_census.describe()

,age,final-weight,education-num,capital-gain,capital-loos,hour-per-week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


## Caracteristicas da Base
- População amostral: 32.561
- Minorias importantes: 28 e 90, vol. Ed.=9
e 16,
ht= 40 e 99
- Media amostral: idade = 38.58, Ed = 10,
Gain. 1077.65, Loss = 87.30, ht= 40.44
- Desvio Padrão: [limites]: idade = 24.94, Ed. = 7.
51, Gain. = -6757.64 Loss= -315.66, ht= 28,9

In [55]:
import plotly.express as px
import pandas as pd

# Usando o caminho do arquivo carregado localmente no Colab
base_census = pd.read_csv('/content/census.csv')
grafico_classe_wa = px.treemap(base_census, path=['workclass', 'age'])
grafico_classe_wa.show()

In [56]:
grafico_classe_ora = px.treemap(base_census_limpa, path=['occupation', 'relationship', 'age'])
grafico_classe_ora.show()

In [57]:
grafico_fineture_eri = px.parallel_categories(base_census, dimensions= ['education','race','income'])
grafico_fineture_eri.show()

# Analise de Anomalias de ddos [Censo]

# Analilise de Anomalias de Dados [Census]


**Enunciado: Auditoria e Higienização de Dados Censitários**

Vamos analizar e estruturar a ingestão de um novo pipeline de dados sociodemográficos focado no desenvolvimento de modelos de inteligência artificial. O arquivo **census.csv** foi extraído de um sistema legado, mas o processo automatizado falhou antes mesmo da etapa de modelagem estatística. Como futuros cientistas da computação e engenheiros de dados, a missão de vocês é realizar uma auditoria rigorosa neste arquivo para garantir a integridade da informação.

Dicas de Investigação e Solução

* **Identificação de Linhas Infiltradas**: Ao forçarem o carregamento inicial (pulando as linhas defeituosas do topo), observem o comportamento dos tipos de dados informados pelo `.dtypes`. Se colunas que deveriam ser numéricas (`age` ou `hour-per-week`) foram lidas como `object`, existem *strings* perdidas no meio da massa de dados. O cabeçalho e fragmentos de texto corrompido estão repetidos ao longo do arquivo.
* **Inconsistência Categórica e Caracteres de Escape:** Analisem a distribuição de classes na coluna `income` através do `.value_counts()`. A mesma classe de renda foi acidentalmente dividida em múltiplas categorias distintas devido a falhas de exportação com caracteres de escape textuais (como `\<=50K` versus `<=50K`).
* **Dados Faltantes Não Convencionais:** Avaliem como a ausência de informação foi mascarada pelo sistema de origem. Em bases censitárias clássicas, campos em branco (como `workclass` e `native-country`) frequentemente não aparecem como `NaN` estruturados, mas são substituídos por marcadores categóricos arbitrários, como o caractere `?`.
* **Tratamento de Espaçamento:** A separação por vírgulas pode conter espaços em branco residuais no início das palavras, prejudicando filtros e agrupamentos futuros. O parâmetro `skipinitialspace=True` na leitura do Pandas é um atalho eficiente para padronizar as bordas do texto.

In [61]:
base_census[(base_census['age'] <= 24) & (base_census.isin([' ?']).any(axis=1))]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
51,18,Private,226956,HS-grad,9,Never-married,Other-service,Own-child,White,Female,0,0,30,?,<=50K
106,17,?,304873,10th,6,Never-married,?,Own-child,White,Female,34095,0,32,United-States,<=50K
201,19,?,170653,HS-grad,9,Never-married,?,Own-child,White,Male,0,0,40,Italy,<=50K
243,20,?,119156,Some-college,10,Never-married,?,Own-child,White,Male,0,0,20,United-States,<=50K
266,19,?,218956,Some-college,10,Never-married,?,Own-child,White,Male,0,0,24,Canada,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32275,20,?,99891,Some-college,10,Never-married,?,Own-child,White,Female,0,0,30,United-States,<=50K
32310,21,?,205939,Some-college,10,Never-married,?,Own-child,White,Male,0,0,40,United-States,<=50K
32314,18,Never-worked,153663,Some-college,10,Never-married,?,Own-child,White,Male,0,0,4,United-States,<=50K
32335,18,?,156608,11th,7,Never-married,?,Own-child,White,Female,0,0,25,United-States,<=50K


In [64]:
def normalizarEstudantes(df):
  condition_age_24 = df['age']<24
  df.loc[condition_age_24 & (df['workclass']==' ?'), 'workclass'] = 'Dependent'
  df.loc[condition_age_24 & (df['ocupation']==' ?'), 'ocupation'] = 'Student'
  return df
base_census = normalizarEstudantes(base_census)

KeyError: 'ocupation'

In [63]:
base_census.loc[base_census['age'] < 18]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
106,17,?,304873,10th,6,Never-married,?,Own-child,White,Female,34095,0,32,United-States,<=50K
209,17,Private,65368,11th,7,Never-married,Sales,Own-child,White,Female,0,0,12,United-States,<=50K
262,17,Private,245918,11th,7,Never-married,Other-service,Own-child,White,Male,0,0,12,United-States,<=50K
271,17,Private,191260,9th,5,Never-married,Other-service,Own-child,White,Male,1055,0,24,United-States,<=50K
335,17,Private,270942,5th-6th,3,Never-married,Other-service,Other-relative,White,Male,0,0,48,Mexico,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
31772,17,?,256173,10th,6,Never-married,?,Own-child,White,Female,0,0,15,United-States,<=50K
31864,17,Private,214787,12th,8,Never-married,Adm-clerical,Own-child,White,Female,0,0,25,United-States,<=50K
31959,17,Private,24090,HS-grad,9,Never-married,Exec-managerial,Own-child,White,Female,0,0,35,United-States,<=50K
32282,17,Private,311907,11th,7,Never-married,Other-service,Own-child,White,Male,0,0,25,United-States,<=50K
